File: strata-rs/src/decode.rs
use crate::error::{DecodeError, DecodeErrorKind};
use crate::value::Value;

pub struct Decoder<'a> {
    input: &'a [u8],
    offset: usize,
}

impl<'a> Decoder<'a> {
    fn new(input: &'a [u8]) -> Self {
        Self { input, offset: 0 }
    }

    fn err(&self, kind: DecodeErrorKind) -> DecodeError {
        DecodeError {
            kind,
            offset: self.offset,
        }
    }

    fn remaining(&self) -> usize {
        self.input.len() - self.offset
    }

    fn read_byte(&mut self) -> Result<u8, DecodeError> {
        if self.offset >= self.input.len() {
            return Err(self.err(DecodeErrorKind::UnexpectedEOF));
        }
        let byte = self.input[self.offset];
        self.offset += 1;
        Ok(byte)
    }

    fn read_slice(&mut self, len: usize) -> Result<&'a [u8], DecodeError> {
        if self.remaining() < len {
            return Err(self.err(DecodeErrorKind::UnexpectedEOF));
        }
        let slice = &self.input[self.offset..self.offset + len];
        self.offset += len;
        Ok(slice)
    }

    fn decode_uleb128(&mut self) -> Result<u64, DecodeError> {
        let mut result = 0u64;
        let mut shift = 0;

        loop {
            let byte = self.read_byte()?;
            let value = (byte & 0x7F) as u64;

            if shift >= 64 {
                return Err(self.err(DecodeErrorKind::InvalidVarint));
            }

            result |= value << shift;

            if (byte & 0x80) == 0 {
                return Ok(result);
            }

            shift += 7;
        }
    }

    fn decode_sleb128(&mut self) -> Result<i64, DecodeError> {
        let mut result = 0i64;
        let mut shift = 0;
        let mut byte;

        loop {
            byte = self.read_byte()? as i64;
            let value = byte & 0x7F;

            result |= value << shift;
            shift += 7;

            if (byte & 0x80) == 0 {
                break;
            }

            if shift >= 64 {
                return Err(self.err(DecodeErrorKind::InvalidVarint));
            }
        }

        if (shift < 64) && ((byte & 0x40) != 0) {
            result |= (!0i64) << shift;
        }

        Ok(result)
    }

    pub fn decode_value(&mut self) -> Result<Value, DecodeError> {
        let tag = self.read_byte()?;

        match tag {
            0x00 => Ok(Value::Null),
            0x01 => Ok(Value::Bool(false)),
            0x02 => Ok(Value::Bool(true)),

            0x10 => {
                let n = self.decode_sleb128()?;
                Ok(Value::Int(n))
            }

            0x20 => {
                let len = self.decode_uleb128()? as usize;
                let start = self.offset;
                let bytes = self.read_slice(len)?;
                let s = std::str::from_utf8(bytes).map_err(|_| DecodeError {
                    kind: DecodeErrorKind::InvalidUtf8,
                    offset: start,
                })?;

                Ok(Value::String(s.to_string()))
            }

            0x21 => {
                let len = self.decode_uleb128()? as usize;
                let bytes = self.read_slice(len)?;
                Ok(Value::Bytes(bytes.to_vec()))
            }

            0x30 => {
                let count = self.decode_uleb128()? as usize;
                let mut items = Vec::with_capacity(count);
                for _ in 0..count {
                    items.push(self.decode_value()?);
                }
                Ok(Value::List(items))
            }

            0x40 => {
                let count = self.decode_uleb128()? as usize;
                let mut map = std::collections::BTreeMap::new();

                for _ in 0..count {
                    let key = match self.decode_value()? {
                        Value::String(s) => s,
                        _ => return Err(self.err(DecodeErrorKind::InvalidTag(tag))),
                    };
                    let value = self.decode_value()?;
                    map.insert(key, value);
                }
                Ok(Value::Map(map))
            }

            other => Err(self.err(DecodeErrorKind::InvalidTag(other))),
        }
    }
}

pub fn decode(input: &[u8]) -> Result<Value, DecodeError> {
    let mut decoder = Decoder::new(input);
    let value = decoder.decode_value()?;

    if decoder.remaining() != 0 {
        return Err(DecodeError {
            kind: DecodeErrorKind::TrailingBytes,
            offset: decoder.offset,
        });
    }

    Ok(value)
}

File: strata-js/src/decode.ts
import { Value } from "./value.js";
import { V } from "./value_factory.js";
import { decodeULEB128, decodeSLEB128 } from "./varint.js";
import { DecodeError } from "./decode_error.js";

const td = new TextDecoder("utf-8", { fatal: true });

const TAG_NULL = 0x00;
const TAG_FALSE = 0x01;
const TAG_TRUE = 0x02;

const TAG_INT = 0x10;
const TAG_STRING = 0x20;
const TAG_BYTES = 0x21;

const TAG_LIST = 0x30;
const TAG_MAP = 0x40;

class Decoder {
    private offset = 0;

    constructor(private readonly input: Uint8Array) { }

    private remaining(): number {
        return this.input.length - this.offset;
    }

    private readByte(): number {
        if (this.offset >= this.input.length) {
            throw new DecodeError({ type: "UnexpectedEOF" }, this.offset);
        }

        return this.input[this.offset++]!;
    }

    private readSlice(len: number): Uint8Array {
        if (this.remaining() < len) {
            throw new DecodeError({ type: "UnexpectedEOF" }, this.offset);
        }

        const slice = this.input.subarray(this.offset, this.offset + len);
        this.offset += len;
        return slice;
    }

    decodeValue(): Value {
        const tagOffset = this.offset;
        const tag = this.readByte();

        switch (tag) {
            case TAG_NULL:
                return V.null();

            case TAG_FALSE:
                return V.bool(false);

            case TAG_TRUE:
                return V.bool(true);

            case TAG_INT: {
                try {
                    const { value, nextOffset } = decodeSLEB128(this.input, this.offset);
                    this.offset = nextOffset;
                    return V.int(value);
                } catch {
                    throw new DecodeError({ type: "InvalidVarint" }, tagOffset);
                }
            }

            case TAG_STRING: {
                let len: bigint;
                try {
                    const result = decodeULEB128(this.input, this.offset);
                    len = result.value;
                    this.offset = result.nextOffset;
                } catch {
                    throw new DecodeError({ type: "InvalidVarint" }, tagOffset);
                }

                const start = this.offset;
                const bytes = this.readSlice(Number(len));

                try {
                    const str = td.decode(bytes);
                    return V.string(str);
                } catch {
                    throw new DecodeError({ type: "InvalidUTF8" }, start);
                }
            }

            case TAG_BYTES: {
                let len: bigint;
                try {
                    const result = decodeULEB128(this.input, this.offset);
                    len = result.value;
                    this.offset = result.nextOffset;
                } catch {
                    throw new DecodeError({ type: "InvalidVarint" }, tagOffset);
                }

                const bytes = this.readSlice(Number(len));
                return V.bytes(bytes);
            }

            case TAG_LIST: {
                let count: bigint;

                try {
                    const result = decodeULEB128(this.input, this.offset);
                    count = result.value;
                    this.offset = result.nextOffset;
                } catch {
                    throw new DecodeError({ type: "InvalidVarint" }, tagOffset);
                }

                const items: Value[] = [];
                for (let i = 0n; i < count; i++) {
                    items.push(this.decodeValue());
                }

                return V.list(items);
            }

            case TAG_MAP: {
                let count: bigint;
                try {
                    const result = decodeULEB128(this.input, this.offset);
                    count = result.value;
                    this.offset = result.nextOffset;
                } catch {
                    throw new DecodeError({ type: "InvalidVarint" }, tagOffset);
                }

                const map = new Map<string, Value>();

                for (let i = 0n; i < count; i++) {
                    const keyValue = this.decodeValue();
                    if (keyValue.kind !== "string") {
                        throw new DecodeError({ type: "InvalidTag", tag }, this.offset);
                    }

                    const value = this.decodeValue();
                    map.set(keyValue.value, value);
                }

                return V.map(map);
            }

            default:
                throw new DecodeError({ type: "InvalidTag", tag }, tagOffset);
        }
    }
}

export function decodeValue(input: Uint8Array): Value {
    const decoder = new Decoder(input);
    const value = decoder.decodeValue();

    if (decoder["remaining"]() !== 0) {
        throw new DecodeError(
            { type: "TrailingBytes" },
            input.length - decoder["remaining"](),
        );
    }

    return value;
}
